lib/nixos-system.nix

```diff
@@ -1,43 +1,93 @@
 { self, lib, home-manager }:
 
-{ modules ? [ ], globalSpecialArgs ? { }, nixosSpecialArgs ? { }, hmSpecialArgs ? { }, ... } @ args:
-lib.nixosSystem (lib.recursiveUpdate (removeAttrs args [ "globalSpecialArgs" "nixosSpecialArgs" "hmSpecialArgs" ]) {
+with lib;
+
+{
+  # The global modules are included in both NixOS and home-manager.
+  globalModules ? [ ]
+
+  # Home-manager specific modules.
+, hmModules ? [ ]
+
+  # NixOS specific modules.
+, nixosModules ? [ ]
+
+  # The global extra arguments are included in both NixOS and home-manager.
+, globalSpecialArgs ? { }
+
+  # Home-manager specific extra arguments.
+, hmSpecialArgs ? { }
+
+  # NixOS specific extra arguments.
+, nixosSpecialArgs ? { }
+
+, ...
+} @ args:
+let
+  args' = removeAttrs args [
+    "globalModules"
+    "nixosModules"
+    "hmModules"
+
+    "globalSpecialArgs"
+    "nixosSpecialArgs"
+    "hmSpecialArgs"
+  ];
+in
+nixosSystem (recursiveUpdate args' {
   specialArgs = {
-    mode = "NixOS";
-    soxin = self;
-  } // globalSpecialArgs // nixosSpecialArgs;
+    # send home-manager down to the NixOS modules
+    inherit home-manager;
```

Comment:
Why is this required? i.e. why pass it as a specialArg instead of `_module.args` as it's currently done?

Reply:
It works better. This solved one infinite recursion I was having while iterating on this.